app/scripts/controllers/network/network-controller.test.js

```diff
@@ -50,224 +77,4799 @@ const PRE_1559_BLOCK = {
+                  },
+                },
+              },
+              async ({ controller, network }) => {
+                network.mockEssentialRpcCalls({
+                  eth_getBlockByNumber: {
+                    response: {
+                      error: 'oops',
+                    },
+                  },
+                });
+                await withoutCallingGetEIP1559Compatibility({
+                  controller,
+                  operation: async () => {
+                    await controller.initializeProvider();
+                  },
+                });
+                expect(
+                  controller.store.getState().networkDetails.EIPS['1559'],
+                ).toBeUndefined();
+
+                await waitForStateChanges({
+                  controller,
+                  propertyPath: ['networkDetails'],
+                  count: 0,
+                  operation: async () => {
+                    await controller.getEIP1559Compatibility();
+                  },
+                });
+                expect(
+                  controller.store.getState().networkDetails.EIPS['1559'],
+                ).toBeUndefined();
+              },
+            );
+          });
+        });
+
+        describe('if the network was switched after the net_version request started but before it completed', () => {
+          it(`persists to state the network version of the newly switched network, not the initial one for ${networkName}`, async () => {
+            const oldNetworkVersion = networkVersion;
+            const newNetworkName = 'goerli';
+            const newNetworkVersion = '5';
+
+            await withController(
+              {
+                state: {
+                  provider: {
+                    type: networkType,
+                  },
+                },
+              },
+              async ({ controller }) => {
+                let netVersionCallCount = 0;
+
+                const fakeProviders = [
+                  {
+                    sendAsync(request, callback) {
+                      if (request.method === 'net_version') {
+                        netVersionCallCount += 1;
+                        if (netVersionCallCount === 1) {
+                          waitForStateChanges({
+                            controller,
+                            propertyPath: ['network'],
+                            operation: () => {
+                              controller.setProviderType(newNetworkName);
+                            },
+                          })
+                            .then(() => {
+                              callback(null, {
+                                id: request.id,
+                                jsonrpc: '2.0',
+                                result: oldNetworkVersion,
+                              });
+                            })
+                            .catch((error) => {
+                              throw error;
+                            });
```

Comment:
Nit: We can get rid of this `.then` while preserving this as a non-blocking operation by using an inline function:

```suggestion
                          const returnOldNetworkVersionAfterStateChange = async () => {
                            await waitForStateChanges({
                              controller,
                              propertyPath: ['network'],
                              operation: () => {
                                controller.setProviderType(newNetworkName);
                              },
                            });
                            callback(null, {
                              id: request.id,
                              jsonrpc: '2.0',
                              result: oldNetworkVersion,
                            });
                          };
                          
                          // intentionally not `await`-ing this because `sendAsync` does not return a Promise
                          returnOldNetworkVersionAfterStateChange();
```

Reply:
Oh, I see the same applies for the next test too, I think that's the last one though

Reply:
I think I might leave this one alone if you don't mind. Since `sendAsync` doesn't return a promise, using the `.then` pattern here seems less surprising than using the `async` keyword. And then we also don't have to explain why we're not `await`ing such a function.

Reply:
Ok, sounds good. Personally I don't find `.then` less surprising, but I see what you mean.

I made this suggestion because I'm anticipating we'll be bringing in the lint rule to prefer async/await, and because it has better stack traces. But neither is important here. Maybe this test will be obsolete by that point anyway.